hybrid/flow.py

```diff
@@ -745,51 +754,58 @@ def iteration_update(self, iterno, cnt, input_state, output_state):
 
     def next(self, state):
         iterno = 0
-        cnt = self.convergence
+        cnt = self.convergence or 0
         input_state = state
+        output_state = input_state
 
-        while True:
+        while not self._stop_event.is_set():
             output_state = self.runnable.run(input_state, executor=immediate_executor).result()
 
             iterno, cnt, input_state = self.iteration_update(iterno, cnt, input_state, output_state)
 
             if self.max_iter is not None and iterno >= self.max_iter:
                 break
-            if cnt <= 0:
+            if self.convergence is not None and cnt <= 0:
                 break
 
         return output_state
 
-
-class LoopN(LoopUntilNoImprovement):
-    """Iterate `runnable` exactly `n` times, using each output as input in the
-    next iteration.
-    """
-
-    def __init__(self, runnable, n):
-        super(LoopN, self).__init__(
-            runnable=runnable, max_iter=n, convergence=n, key=lambda _: 0)
+    def halt(self):
+        self.runnable.stop()
+        self._stop_event.set()
 
 
 class Loop(LoopUntilNoImprovement):
     pass
 
+
 SimpleIterator = Loop
 
 
 class LoopWhileNoImprovement(LoopUntilNoImprovement):
     """Iterates `runnable` until a state quality metric, defined by the `key`
     function, shows no improvement for at least `max_tries` number of
-    iterations.
+    iterations (or until `max_iter` number of iterations is exceeded).
+
+    Note:
+        Unlike `LoopUntilNoImprovement`/`Loop`, `LoopWhileNoImprovement` will
+        run the loop body runnable with the same input if it shows no
+        improvement, and it will use the new output if it's better than previous.
 
     Args:
         runnable (:class:`Runnable`):
             A runnable that's looped over.
 
-        max_tries (int, optional, default=10):
+        max_iter (int/None, optional, default=None):
+            Maximum number of times the `runnable` is run, regardless of other
+            termination criteria. This is the upper bound. If set to `None`,
+            upper bound on the number or iterations is not set (the default).
```

Comment:
"By default, no upper bound is set..." Also, I think you mean "number of iterations" instead of "number or iterations"

Reply:
Fixed, thanks.